### ISSUE TITLE
[FEA] Investigate if `release_assert` impacts performance

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#3044]
**Is your feature request related to a problem? Please describe.**
[`release_assert`](https://github.com/rapidsai/cudf/blob/66aeacf4de5816eada4dea45fb9f2cf624e31ba2/cpp/src/utilities/release_assert.cuh) is used throughout libcudf code to signal an error in device code. 

Most concerning is that this is used in type_dispatcher error path.

It's come to my attention that using the `__assertfail` may incur syscalls, which can force ABI constraints that result in a performance impact even in the happy path. 

**Describe the solution you'd like**

We should benchmark with and without the `release_assert` in the type dispatcher (using the type dispatcher benchmark to start) and see if there is any difference in performance.

**Comment by jrhemstad** [comment]
https://github.com/rapidsai/cudf/pull/7583

Turns out, it does! Kind of. It has a pretty sizable impact on binary size and compile time. Actual runtime performance impact seemed more minimal.

**Comment by harrism** [comment]
Should we close this now?

**Comment by jrhemstad** [comment]
Yeah, I think it's fine to close this now given that #7583 made it so the assert is no longer present in Release code.
